### PR TITLE
Implement sub group reduce

### PR DESCRIPTION
**Pull request by AidanBeltonS**
This PR adds support for `reduce_over_sub_group`. It is wired into the existing reduce and can be enabled with `USE_SUB_GROUP_REDUCE`

### CODE REVIEW

File: include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h [hunk 1]
```diff
@@ -21,6 +21,7 @@
 
 #include "../../onedpl_config.h"
 #include "../../utils.h"
+#include "../../iterator_impl.h"
```
Comment (joeatodd) [inline include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h:24]: Is this include really needed?

**Review by joeatodd — APPROVED** [review]
LGTM - couple of minor comments.

File: include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h (hunk 1 above)
Comment (AidanBeltonS) [reply include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h]: Removed

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h [hunk 2]
```diff
@@ -46,6 +46,29 @@ template <typename... _Name>
 class __reduce_kernel;
 
 
+// Single work group kernel that transforms and reduces __n elements to the single result.
+template <typename _Tp, typename _NDItemId, typename _Size, typename _TransformPattern, typename _ReducePattern,
+          typename _InitType, typename _AccLocal, typename _Res, typename... _Acc>
+inline void
+__sub_group_reduce_kernel(const _NDItemId& __item_id, const _Size& __n, _TransformPattern __transform_pattern,
```
[the remark this answers is not in the document]
Comment (AidanBeltonS) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h]: Removed the duplicate function

**Pull request merged**